File: ios/Cove/SettingsScreen/ChangePinView.swift
//
//  ChangePinView.swift
//  Cove
//
//  Created by Praveen Perera on 12/12/24.
//

import SwiftUI

private enum PinState {
    case current, new, confirm(String)
}

struct ChangePinView: View {
    /// args
    var isPinCorrect: (String) -> Bool
    var backAction: () -> Void
    var onComplete: (String) -> Void

    /// private
    @State private var pinState: PinState = .current

    var body: some View {
        Group {
            switch pinState {
            case .current:
                NumberPadPinView(
                    title: "Enter Current PIN",
                    isPinCorrect: isPinCorrect,
                    backAction: backAction,
                    onUnlock: { _ in
                        withAnimation {
                            pinState = .new
                        }
                    }
                )

            case .new:
                NumberPadPinView(
                    title: "Enter new PIN",
                    isPinCorrect: { _ in true },
                    showPin: false,
                    backAction: backAction,
                    onUnlock: { enteredPin in
                        withAnimation {
                            pinState = .confirm(enteredPin)
                        }
                    }
                )

            case let .confirm(pinToConfirm):
                NumberPadPinView(
                    title: "Confirm New PIN",
                    isPinCorrect: { $0 == pinToConfirm },
                    showPin: false,
                    backAction: backAction,
                    onUnlock: onComplete
                )
            }
        }
    }
}

#Preview {
    ChangePinView(
        isPinCorrect: { pin in pin == "111111" },
        backAction: {},
        onComplete: { _ in }
    )
}

File: android/app/src/main/java/org/bitcoinppl/cove/flows/SettingsFlow/ChangePinView.kt
package org.bitcoinppl.cove.flows.SettingsFlow

import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.res.stringResource
import org.bitcoinppl.cove.R
import org.bitcoinppl.cove.views.NumberPadPinView

private sealed class ChangePinState {
    data object Current : ChangePinState()

    data object New : ChangePinState()

    data class Confirm(
        val pinToConfirm: String,
    ) : ChangePinState()
}

@Composable
fun ChangePinView(
    isPinCorrect: (String) -> Boolean,
    backAction: () -> Unit,
    onComplete: (String) -> Unit,
) {
    var pinState: ChangePinState by remember { mutableStateOf(ChangePinState.Current) }

    when (val state = pinState) {
        is ChangePinState.Current -> {
            NumberPadPinView(
                title = stringResource(R.string.settings_pin_enter_current),
                isPinCorrect = isPinCorrect,
                backAction = backAction,
                onUnlock = {
                    pinState = ChangePinState.New
                },
            )
        }

        is ChangePinState.New -> {
            NumberPadPinView(
                title = stringResource(R.string.settings_pin_enter_new),
                isPinCorrect = { true },
                backAction = backAction,
                onUnlock = { enteredPin ->
                    pinState = ChangePinState.Confirm(enteredPin)
                },
            )
        }

        is ChangePinState.Confirm -> {
            NumberPadPinView(
                title = stringResource(R.string.settings_pin_confirm_new),
                isPinCorrect = { it == state.pinToConfirm },
                backAction = backAction,
                onUnlock = onComplete,
            )
        }
    }
}
